include/deal.II/numerics/smoothness_estimator.templates.h

```diff
@@ -0,0 +1,434 @@
+     * $0 < i*i+j*j < N*N$ and $0 < i*i+j*j+k*k < N*N$ in 2D and 3D,
+     * respectively.
+     */
+    template <int dim>
+    std::pair<bool, unsigned int>
+    predicate(const TableIndices<dim> &ind, const unsigned int max_degree)
+    {
+      unsigned int v = 0;
+      for (unsigned int i = 0; i < dim; i++)
+        v += ind[i] * ind[i];
+      if (v > 0 && v < max_degree * max_degree)
+        return std::make_pair(true, v);
+      else
+        return std::make_pair(false, v);
+    }
+  } // namespace
+
+
+  template <int dim, int spacedim, typename VectorType>
+  void
+  legendre_coeff_decay(
+    FESeries::Legendre<dim, spacedim> &    fe_legendre,
+    const hp::DoFHandler<dim, spacedim> &  dof_handler,
+    const std::vector<const VectorType *> &all_solutions,
+    const std::vector<Vector<float> *> &   all_smoothness_indicators,
+    const double                           smallest_abs_coefficient)
+  {
+    Assert(smallest_abs_coefficient >= 0.,
+           ExcMessage("smallest_abs_coefficient should be non-negative."));
+
+    using number = typename VectorType::value_type;
+    using number_coeff =
+      typename FESeries::Legendre<dim, spacedim>::CoefficientType;
+
+    AssertDimension(all_solutions.size(), all_smoothness_indicators.size());
+
+    for (auto &smoothness_indicator : all_smoothness_indicators)
+      smoothness_indicator->reinit(
+        dof_handler.get_triangulation().n_active_cells());
+
+    Table<dim, number_coeff> expansion_coefficients;
+    resize(expansion_coefficients, fe_legendre.get_size_in_each_direction());
+
+    Vector<number> local_dof_values;
+
+    // auxiliary vector to do linear regression
+    std::vector<number_coeff> x;
+    std::vector<number_coeff> y;
+
+    x.reserve(dof_handler.get_fe_collection().max_degree());
+    y.reserve(dof_handler.get_fe_collection().max_degree());
+
+    for (auto &cell : dof_handler.active_cell_iterators())
+      if (cell->is_locally_owned())
+        {
+          local_dof_values.reinit(cell->get_fe().dofs_per_cell);
+
+          const unsigned int pe = cell->get_fe().degree;
+
+          // since we use coefficients with indices [1,pe] in each direction,
+          // the number of coefficients we need to calculate is at least N=pe+1
+          AssertIndexRange(pe, fe_legendre.get_size_in_each_direction());
+
+          auto solution_it              = all_solutions.cbegin();
+          auto smoothness_indicators_it = all_smoothness_indicators.begin();
+          for (; solution_it != all_solutions.cend();
+               ++solution_it, ++smoothness_indicators_it)
+            {
+              // for linear elements set analyticity to 0 (i.e. smooth)
+              // as we can't really estimate anything reasonably
+              if (pe == 1)
+                {
+                  (*(*smoothness_indicators_it))(cell->active_cell_index()) = 0;
+                  continue;
+                }
```

Comment:
I think you could just move that `if (pe==1)` outside the loop over all vectors. As in
```
  if (pe == 1)
  {
    for (auto smoothness_indicators_it : ...)
      for (auto x : *smoothness_indicators_it)
        x = 0;
  }
  else
    // actually do something here
```
This might make it easier to read.

Reply:
personally, I think it would be harder to read. If I have that corner case, i would still need to loop over all output vectors to set values accordingly so I would end up writing `for` loop twice.